### DISCUSSION TITLE
Websocket Disconnect Potentially Broken

### DISCUSSION THREAD

**Discussion by econeale** [fastapi/fastapi#11008] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import asyncio
import logging

import fastapi
import uvicorn

APP = fastapi.FastAPI()


@APP.websocket("/ws")
async def ws_telemetry(websocket: fastapi.WebSocket) -> None:
    await websocket.accept()

    try:
        while True:
            print("Still Looping...")
            await asyncio.sleep(1)

    except fastapi.WebSocketDisconnect:
        logging.info("Websocket disconnected")


if __name__ == "__main__":
    uvicorn.run("websocket_duplicator:APP", host="127.0.0.1", port=8080, reload=True)
```

### Description

When running FastAPI with a websocket, the WebSocketDisconnect error isn't raised per the [documentation](https://fastapi.tiangolo.com/advanced/websockets/#handling-disconnections-and-multiple-clients). Using the given reproducer code and Postman's websocket function, I get the following output:

```
>$ python3 ./websocket_duplicator.py
INFO:     Will watch for changes in these directories: ['/home/neale/btbr-middleware']
INFO:     Uvicorn running on http://127.0.0.1:8080 (Press CTRL+C to quit)
INFO:     Started reloader process [3256733] using WatchFiles
INFO:     Started server process [3256735]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     ('127.0.0.1', 42336) - "WebSocket /ws" [accepted]
Still Looping...
INFO:     connection open
Still Looping...
Still Looping...
Still Looping...
INFO:     connection closed
Still Looping...
Still Looping...
Still Looping...
Still Looping...
Still Looping...
Still Looping...
```

It seems to me that this functionality has broken as similar code worked with FastAPI 0.95.2 however, I found a [Starlette discussion ](https://github.com/encode/starlette/issues/759) which makes me think the behavior is, for some reason, intentional. My question is: is the WebSocketDisconnect exception supposed to be raised when a client disconnects or is the feature broken? 

### Operating System

Linux

### FastAPI Version

0.109.0

### Pydantic Version

2.5.3

### Python Version

3.10

### Additional Context

starlette==0.35.1
uvicorn==0.26.0
websockets==12.0

**Comment by Kludex** [comment, marked as answer]
Your snippet is unrealistic. The `WebSocketDisconnect` only raises on `websocket.receive()`.

Note: the next release of Uvicorn/Starlette will raise `WebSocketDisconnect` on `send()` as well.

**Reply by econeale** [reply]
I'm not sure what you mean by unrealistic. Sending status information from a server to a client seems like a pretty standard use case. For example, here's a function that worked with FastAPI 0.95.2 and no longer works in version 0.109.0. It seems silly to have to implement a separate heartbeat function in the application logic when the keepalive is already implemented in the websocket logic.

```
@APP.websocket("/get_serial_state")
async def get_serial_state(websocket: fastapi.WebSocket):
    await websocket.accept()
    to_send = {}
    try:
        while True:
            to_send["state"] = SERIAL_PROCESS_OBJ.serial_ok()
            await websocket.send_json(to_send)
            await asyncio.sleep(SOCKET_UPDATE_SECONDS)
    except fastapi.WebSocketDisconnect:
        pass
    except fastapi.WebSocketException:
        pass
```

**Reply by Kludex** [reply]
Your `while True` above only had a sleep...

As I've said, `receive()` and `send()` (from the latest Starlette) raise `WebSocketDisconnect`.

**Reply by econeale** [reply]
Well then I'll wait for FastAPI to update its dependencies and see if it fixes my issue. Thank you.

**Reply by Kludex** [reply]
> > Your `while True` above only had a sleep...
> > As I've said, `receive()` and `send()` (from the latest Starlette) raise `WebSocketDisconnect`.
> 
> Thanks! It solved my Issue. how do you know that? can you give me the ref?

Because I'm a Starlette maintainer. :trollface: 

Yes, I can: https://github.com/encode/starlette/releases/tag/0.36.0

**Reply by vatheara** [reply]
I just deleted my previous comment because I don't want to bother you. you had me everything when I go to your profile 🫡.
Anyway thank for your instant notice! 
Appreciate your works!
